sklearn/model_selection/tests/test_split.py

```diff
@@ -0,0 +1,959 @@
+
+    # Check that no label is on both sides of the split
+    labels = np.asarray(labels, dtype=object)
+    for train, test in lkf.split(X, y, labels):
+        assert_equal(len(np.intersect1d(labels[train], labels[test])), 0)
+
+    # Construct the test data
+    labels = ['Albert', 'Jean', 'Bertrand', 'Michel', 'Jean',
+              'Francis', 'Robert', 'Michel', 'Rachel', 'Lois',
+              'Michelle', 'Bernard', 'Marion', 'Laura', 'Jean',
+              'Rachel', 'Franck', 'John', 'Gael', 'Anna', 'Alix',
+              'Robert', 'Marion', 'David', 'Tony', 'Abel', 'Becky',
+              'Madmood', 'Cary', 'Mary', 'Alexandre', 'David', 'Francis',
+              'Barack', 'Abdoul', 'Rasha', 'Xi', 'Silvia']
+
+    n_labels = len(np.unique(labels))
+    n_samples = len(labels)
+    n_folds = 5
+    tolerance = 0.05 * n_samples  # 5 percent error allowed
+    ideal_n_labels_per_fold = n_samples // n_folds
+
+    X = y = np.ones(n_samples)
+
+    # Get the test fold indices from the test set indices of each fold
+    folds = np.zeros(n_samples)
+    for i, (_, test) in enumerate(lkf.split(X, y, labels)):
+        folds[test] = i
+
+    # Check that folds have approximately the same size
+    assert_equal(len(folds), len(labels))
+    for i in np.unique(folds):
+        assert_greater_equal(tolerance,
+                             abs(sum(folds == i) - ideal_n_labels_per_fold))
+
+    # Check that each label appears only in 1 fold
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", DeprecationWarning)
+        for label in np.unique(labels):
+            assert_equal(len(np.unique(folds[labels == label])), 1)
+
+    # Check that no label is on both sides of the split
+    labels = np.asarray(labels, dtype=object)
+    for train, test in lkf.split(X, y, labels):
+        assert_equal(len(np.intersect1d(labels[train], labels[test])), 0)
+
+    # Should fail if there are more folds than labels
+    labels = np.array([1, 1, 1, 2, 2])
+    X = y = np.ones(len(labels))
+    assert_raises_regexp(ValueError, "Cannot have number of folds.*greater",
+                         next, LabelKFold(n_folds=3).split(X, y, labels))
+
+
+def test_nested_cv():
+    # Test if nested cross validation works with different combinations of cv
+    rng = np.random.RandomState(0)
+
+    X, y = digits.data, digits.target
+    labels = rng.randint(0, 5, len(digits.data))
+
+    cvs = [LeaveOneLabelOut(),
+           LeaveOneOut(),
+           LabelKFold(n_folds=5),
+           StratifiedKFold(n_folds=3),
+           StratifiedShuffleSplit(n_iter=10, random_state=0)]
+
+    for inner_cv, outer_cv in combinations_with_replacement(cvs, 2):
+        print(inner_cv, outer_cv)
+        import time
+        start = time.time()
+        gs = GridSearchCV(LinearSVC(random_state=0), param_grid={'C': [1, 10]},
+                          cv=inner_cv)
+        gs.fit(X, y, labels)
+        cross_val_score(LinearSVC(), X=X, y=y, labels=y, cv=outer_cv)
```

Comment:
needs ``fit_params`` as we discussed. Maybe we want to do the test for all CVS? and use a mock estimator?